proposals/0000-type-vs-constraint.rst

```diff
@@ -0,0 +1,302 @@
+
+  type FUN :: forall (m :: Multiplicity) ->
+              forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+              TYPE r1 -> TYPE r2 -> Type
+
+  -- Data type declarations, used only at the type level
+  data Multiplicity = Many | One
+  data Levity       = Lifted | Unlifted
+  data RuntimeRep   = BoxedRep Levity | IntRep | FloatRep | ...
+
+  -- Type synonyms
+  type LiftedRep   = BoxedRep Lifted
+  type UnliftedRep = BoxedRep Unlifted
+  type Type        = TYPE LiftedRep
+  type (->)        = FUN Many
+
+  -- (=>) is not something that can be written unsaturated;
+  --      rat
+
+NB: in GHC, implicit parameters are internally represented as a special class,
+but that is not user-visible.
+
+Type and Constraint are not apart
+:::::::::::::::::::::::::::::::::::
+
+GHC has an optimization for one-element classes (where the element
+is either a superclass or a method), defining these in like a newtype, not a datatype.
+For example, if we have ::
+
+  class Default a where
+    def :: a
+
+the Core of the program will have a definition like ::
+
+  newtype Default a = MkDefault a
+
+In turn, this newtype gives rise to an axiom (coercion), like so::
+
+  axDefault :: Default a ~R# a
+
+where ``~R#`` represents primitive representational equality. Note that
+``axDefault`` is *heterogeneous*: the kind of ``Default a`` is ``Constraint``,
+whereas the kind of ``a`` is ``Type``.
+
+GHC allows us to extract out an equality relationship between *kinds* from an
+equality relationship on *types* -- and kind equalities are always nominal. To
+wit, Core allows ::
+
+  KindCo axDefault :: Constraint ~# Type
+
+Now, suppose that you could write this::
+
+  type family F a
+  type instance F Type = Int
+  type instance F Constraint = Bool
+
+If these instances were allowed, GHC could
+produce a coercion between ``Int`` and ``Bool``, thus::
+
+  Bool  ~#  F Constraint   -- By type instance F Constraint (backwards)
+        ~#  F Type         -- By KindCo axDefault
+        ~#  Int            -- By type instance F Type
+
+That would be Very, Very Bad.  So, although ``Type`` and ``Constraint`` are built
+with different (un-equal) primitive type constructors,
+
+* **GHC's type checker treats `Type` and `Constraint` as *not apart*.**
+
+That in turn makes GHC complain that the above instances overlap, and are hence illegal.
+
+
+Proposed Change Specification
+-----------------------------
+
+We propose the following new setup, not repeating any types that remains unchanged::
+
+  -- Primitive type constructors
+  type SORT :: TypeOrConstraint -> RuntimeRep -> Type
+  type IP   :: forall (r :: RuntimeRep). Symbol -> TYPE r -> CONSTRAINT r
+
+  type (=>)  :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> TYPE r2 -> Type  -- primitive
+  type (==>) :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> CONSTRAINT r2 -> Constraint
+
+  -- Data types
+  data TypeOrConstraint = TypeLike | ConstraintLike
```

Comment:
Bear in mind that type families cannot appear in instance heads, and that the `TypeLike` data constructor occurs in `Type` by expanding type synonyms. Thus the drawback would be forbidding instances that match on `Type`, which is a pretty major restriction! See discussion starting from https://gitlab.haskell.org/ghc/ghc/-/issues/21623#note_432137

Reply:
I should have added: I think the proposal should somewhere say explicitly that instances matching on `Type` (or `Constraint`) remain allowed, hence matching on `TypeLike` or `ConstraintLike` must be allowed, thus motivating the use of a data kind with non-apart constructors rather than type families.

(I can't help but wonder if there is a connection to homotopy type theory lurking here somewhere, with `TypeOrConstraint` being the type with two points connected by a path.)

Reply:
This makes sense, but I don't understand how we can be allowed to write instance of `TypeLike` and `ConstraintLike` if they are not apart.

Reply:
I see. This is a little unfortunate. But as Simon says: not worse than today.